File: Tutorials/Codes/Modern_Control_Course/Chapter12/Lesson4/Chapter12_Lesson4.m
% Chapter12_Lesson4.m
% Modern Control — Chapter 12, Lesson 4
% Finite-horizon controllability Gramian, energy ellipsoid, and optional Simulink model.
% Requires base MATLAB. Control System Toolbox and Simulink sections are optional.

clear; clc; close all;

A = [-1 0; 0 -4];
B = [1; 0.08];
T = 2.0;
N = 4000;
dt = T/N;

% Wc(T)=int_0^T exp(A*s) B B' exp(A'*s) ds by trapezoidal quadrature.
W = zeros(2,2);
for k = 0:N
    s = k*dt;
    Phi = expm(A*s);
    G = Phi*B*B'*Phi';
    w = 1.0;
    if k == 0 || k == N
        w = 0.5;
    end
    W = W + w*G*dt;
end
W = 0.5*(W + W');

[V,D] = eig(W);
lambda = diag(D);
[lambda,idx] = sort(lambda,'ascend');
V = V(:,idx);

fprintf('Wc(T)=\n'); disp(W);
fprintf('Eigenvalues:\n'); disp(lambda.');
fprintf('Condition number = %.6g\n', lambda(end)/lambda(1));

E = @(delta) delta'*(W\delta);
e1 = [1;0];
e2 = [0;1];
fprintf('Energy to reach e1 = %.8f\n', E(e1));
fprintf('Energy to reach e2 = %.8f\n', E(e2));
fprintf('Best-axis energy = %.8f\n', 1/lambda(end));
fprintf('Worst-axis energy = %.8f\n', 1/lambda(1));

% Plot reachable energy ellipsoid delta' inv(W) delta <= rho^2.
rho = 1;
theta = linspace(0,2*pi,400);
circle = [cos(theta); sin(theta)];
ellipse = V*diag(rho*sqrt(lambda))*circle;
figure;
plot(ellipse(1,:), ellipse(2,:), 'LineWidth', 1.5); grid on; axis equal;
xlabel('state component x_1'); ylabel('state component x_2');
title('Finite-horizon controllability energy ellipsoid');
hold on;
for i = 1:2
    axisVector = V(:,i)*rho*sqrt(lambda(i));
    plot([0 axisVector(1)], [0 axisVector(2)], 'LineWidth', 2);
end

% Optional Control System Toolbox comparison for infinite-horizon stable systems.
if exist('ss','file') == 2 && exist('gram','file') == 2
    sys = ss(A,B,eye(2),zeros(2,1));
    Winf = gram(sys,'c');
    fprintf('Infinite-horizon Wc from Control System Toolbox:\n'); disp(Winf);
end

% Optional Simulink skeleton: create a state-space plant block for simulation.
% Feed a designed input signal u(t) to the input port and inspect state outputs.
if exist('new_system','file') == 2
    model = 'Chapter12_Lesson4_Simulink';
    if bdIsLoaded(model)
        close_system(model,0);
    end
    new_system(model);
    add_block('simulink/Sources/In1', [model '/u']);
    add_block('simulink/Continuous/State-Space', [model '/Plant']);
    add_block('simulink/Sinks/Out1', [model '/x']);
    set_param([model '/Plant'], 'A', mat2str(A), 'B', mat2str(B), ...
        'C', mat2str(eye(2)), 'D', mat2str(zeros(2,1)));
    add_line(model, 'u/1', 'Plant/1');
    add_line(model, 'Plant/1', 'x/1');
    save_system(model);
    fprintf('Created optional Simulink model: %s.slx\n', model);
end

File: Tutorials/Codes/Modern_Control_Course/Chapter12/Lesson4/Chapter12_Lesson4.java
// Chapter12_Lesson4.java
// Modern Control — Chapter 12, Lesson 4
// From-scratch finite-horizon controllability Gramian for a 2-state SISO example.
// No external Java library is required.

public class Chapter12_Lesson4 {
    static double[][] A = {{-1.0, 0.0}, {0.0, -4.0}};
    static double[][] B = {{1.0}, {0.08}};

    static double[][] add(double[][] X, double[][] Y) {
        int r = X.length, c = X[0].length;
        double[][] Z = new double[r][c];
        for (int i = 0; i < r; i++) {
            for (int j = 0; j < c; j++) Z[i][j] = X[i][j] + Y[i][j];
        }
        return Z;
    }

    static double[][] scale(double[][] X, double a) {
        int r = X.length, c = X[0].length;
        double[][] Z = new double[r][c];
        for (int i = 0; i < r; i++) {
            for (int j = 0; j < c; j++) Z[i][j] = a * X[i][j];
        }
        return Z;
    }

    static double[][] mul(double[][] X, double[][] Y) {
        int r = X.length, c = Y[0].length, inner = Y.length;
        double[][] Z = new double[r][c];
        for (int i = 0; i < r; i++) {
            for (int j = 0; j < c; j++) {
                for (int k = 0; k < inner; k++) Z[i][j] += X[i][k] * Y[k][j];
            }
        }
        return Z;
    }

    static double[][] transpose(double[][] X) {
        int r = X.length, c = X[0].length;
        double[][] Z = new double[c][r];
        for (int i = 0; i < r; i++) {
            for (int j = 0; j < c; j++) Z[j][i] = X[i][j];
        }
        return Z;
    }

    static double[][] gramianDerivative(double[][] W) {
        // dW/ds = A W + W A^T + B B^T, W(0)=0.
        return add(add(mul(A, W), mul(W, transpose(A))), mul(B, transpose(B)));
    }

    static double[][] rk4Step(double[][] W, double h) {
        double[][] k1 = gramianDerivative(W);
        double[][] k2 = gramianDerivative(add(W, scale(k1, h / 2.0)));
        double[][] k3 = gramianDerivative(add(W, scale(k2, h / 2.0)));
        double[][] k4 = gramianDerivative(add(W, scale(k3, h)));
        return add(W, scale(add(add(k1, scale(add(k2, k3), 2.0)), k4), h / 6.0));
    }

    static double[][] finiteHorizonGramian(double T, int steps) {
        double h = T / steps;
        double[][] W = {{0.0, 0.0}, {0.0, 0.0}};
        for (int i = 0; i < steps; i++) W = rk4Step(W, h);
        // Symmetrize to remove tiny numerical asymmetry.
        double avg = 0.5 * (W[0][1] + W[1][0]);
        W[0][1] = avg;
        W[1][0] = avg;
        return W;
    }

    static double[] eigenvalues2x2Symmetric(double[][] W) {
        double a = W[0][0], b = W[0][1], d = W[1][1];
        double tr = a + d;
        double disc = Math.sqrt((a - d) * (a - d) + 4.0 * b * b);
        return new double[]{0.5 * (tr - disc), 0.5 * (tr + disc)};
    }

    static double minimumEnergy(double[][] W, double[] delta) {
        double det = W[0][0] * W[1][1] - W[0][1] * W[1][0];
        double inv00 = W[1][1] / det;
        double inv01 = -W[0][1] / det;
        double inv10 = -W[1][0] / det;
        double inv11 = W[0][0] / det;
        return delta[0] * (inv00 * delta[0] + inv01 * delta[1])
             + delta[1] * (inv10 * delta[0] + inv11 * delta[1]);
    }

    static void printMatrix(String name, double[][] X) {
        System.out.println(name + " =");
        for (double[] row : X) {
            for (double v : row) System.out.printf("%14.8f ", v);
            System.out.println();
        }
    }

    public static void main(String[] args) {
        double T = 2.0;
        double[][] W = finiteHorizonGramian(T, 4000);
        printMatrix("Wc(T)", W);
        double[] lambda = eigenvalues2x2Symmetric(W);
        System.out.printf("Eigenvalues: %.10f %.10f%n", lambda[0], lambda[1]);
        System.out.printf("Condition number: %.6f%n", lambda[1] / lambda[0]);
        System.out.printf("Energy to reach e1: %.8f%n", minimumEnergy(W, new double[]{1.0, 0.0}));
        System.out.printf("Energy to reach e2: %.8f%n", minimumEnergy(W, new double[]{0.0, 1.0}));
        System.out.printf("Best-axis energy: %.8f%n", 1.0 / lambda[1]);
        System.out.printf("Worst-axis energy: %.8f%n", 1.0 / lambda[0]);
    }
}
